### PR TITLE
Text overlay visual

### PR DESCRIPTION
**Pull request by lukasz-migas**
* Added new VispyLabelVisual which can be used to display simple text message on the canvas (e.g. fps, slice position or frame number)
* Added new Label model to control the label visual
* Added minimal example to show how the label can be used to display fps

# Description
This pull request adds simple text visual to the canvas to addresses #2572 and [napari/napari-animation#44](https://github.com/napari/napari-animation/issues/44).
The request was for simple text visual in the canvas that can display timestamps (or any simple text) in the canvas without having to add a new layer with TextManager. This minimal change adds new `Label` model to the `ViewerModel` and `VispyLabelVisual` to the `QtViewer` widget. 
The text of the VispyLabelVisual can be updated by simply changing the `viewer.label.text=”Some Text”`. The `Label` model also exposes a couple of attributes such as `color` (could be change to `colored` like in the scalebar), `position` (again, uses the top left, top right, bottom left and bottom right positions, could be adapted to add center top and center bottom) and `font_size`.
 Example:
```
import numpy as np
import napari

def update_fps(fps):
    """Update fps"""
    viewer.label.text = "%1.1f FPS" % fps

viewer = napari.Viewer()
viewer.add_image(np.random.random((5, 5, 5)), colormap='red', opacity=0.8)
viewer.label.visible = True
viewer.window.qt_viewer.canvas.measure_fps(callback=update_fps)

napari.run()
```

## Type of change

- [ ] New feature (non-breaking change which adds functionality)

# References

# How has this been tested?

- [x] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by jni** [comment]
Thanks @lukasz-migas, this is awesome! :tada: 

I won't review yet but can we brainstorm a different name? I feel like "label" is confusing with Labels layers. Some ideas to get started:

viewer.hud
viewer.hud_text
viewer.canvas_text
viewer.text_overlay

Thoughts?

**Comment by lukasz-migas** [comment]
Hi @jni! 

Thanks, I agree, `label`  could be confusing in which case I would say `text_overlay` would be my choice.

**Comment by sofroniewn** [comment]
This is a wonderful contribution @lukasz-migas - it's really perfect, the example works great for me. The test is appropriate, and the doc strings and code look great.

I think my only feedback is that I agree we should change the name from label. I like `text_overlay` best too. If you can make that change then I'd be happy to approve and merge!

**Comment by sofroniewn** [comment]
Here is a screenshot of the example

Screen Shot 2021-04-22 at 8 34 21 PM

File: examples/viewer_fps_label.py
Comment (goanpeca) [inline examples/viewer_fps_label.py]: Docstrings are sentences, so it is generally recommended to use a period at the end.

File: examples/viewer_fps_label.py
Comment (goanpeca) [inline examples/viewer_fps_label.py]: Same here

(Docstrings are sentences, so it is generally recommended to use a period at the end.)

File: examples/viewer_fps_label.py
Comment (goanpeca) [inline examples/viewer_fps_label.py]: Maybe using an f-string would make this more up to date?

```python
viewer.label.text = f"%{fps:1.1f} FPS"
```

**Review by goanpeca — COMMENTED** [review]
This looks great @lukasz-migas 🚀 

Minor (optional) nitpicks :)

**Comment by kevinyamauchi** [comment]
Thanks, @lukasz-migas . As everyone else has said, I think this is amazing. I especially appreciate the test and an example.

I also like the `text_overlay` name.

I am wondering if we want to put this as private or "experimental" for now. I am wondering this because I think it would be nice to consider making the API more general in the not-so-distant future (e.g., allowing multiple text visuals and placement at arbitrary coordinates) and I also think this PR is a great chunk of work and that it is worth merging with its current feature set. Thoughts?

**Comment by lukasz-migas** [comment]
@sofroniewn @goanpeca @kevinyamauchi thanks for the feedback 👍 

I've made the requested changes and also added two extra text positions (top_center, bottom_center) so the `text_overlay` can be used as e.g. title. 

Should the `VispyLabelVisual` be renamed `VispyTextOverlayVisual`?

**Comment by kevinyamauchi** [comment]
Thanks for the updates, @lukasz-migas !

> Should the VispyLabelVisual be renamed VispyTextOverlayVisual?

I agree that we should rename it. What about `VispyTextVisual`?

**Review by goanpeca — APPROVED** [review]

**Comment by sofroniewn** [comment]
> I've made the requested changes and also added two extra text positions (top_center, bottom_center) so the text_overlay can be used as e.g. title.

Love this!

> I am wondering if we want to put this as private or "experimental" for now. I am wondering this because I think it would be nice to consider making the API more general in the not-so-distant future (e.g., allowing multiple text visuals and placement at arbitrary coordinates) and I also think this PR is a great chunk of work and that it is worth merging with its current feature set. Thoughts?

Yeah, I think `viewer.experimental.text_overlay` might be best for now, as I do see the potential for this the API to change in the near term future (but still use a lot of this underlying code). I can right away see people just want to have some text in the top left and some other text in the top right!

**Comment by sofroniewn** [comment]
Maybe @jni can do a final review and weigh in on `experimental` vs not

**Review by sofroniewn — APPROVED** [review]
Ok, thanks for all the re-naming! This looks great @lukasz-migas. I'm approving now and think it's ready to go, but maybe @jni can give it one additional look too before merge.

**Comment by lukasz-migas** [comment]
@sofroniewn should the `color` attribute be changed from `str` type to `ColorType`? (slightly different definition than one used by layers).

```
import numpy as np
from typing import List, Tuple, Union
from vispy.color import Color, ColorArray

ColorType = Union[str, List, Tuple, np.ndarray, Color, ColorArray]
```

**Comment by sofroniewn** [comment]
> @sofroniewn should the color attribute be changed from str type to ColorType? (slightly different definition than one used by layers).

Ah, good catch! We probably want to use our `transform_color` method here. You'll need to add this in a validator - it's a little tricky but you can copy what was done for the `current_color: Optional[Array[float, (4,)]] = None` attribute in the color_manager here

https://github.com/napari/napari/blob/8880d1a368b447d4ebb19444f7ee90f65c70e220/napari/layers/utils/color_manager.py#L147

with this validator https://github.com/napari/napari/blob/8880d1a368b447d4ebb19444f7ee90f65c70e220/napari/layers/utils/color_manager.py#L168-L175

I think it would be good to do this before merge. @kevinyamauchi can also advise if you run into difficulty

File: napari/components/text_overlay.py
Comment (sofroniewn) [inline napari/components/text_overlay.py]: I know I copied and pasted the optional over here, but looking at this again - I think we don't need the optional - there should always be a text color right? We could then drop the return `None` stuff and just have the `transform_color(v)[0]` I think

File: napari/components/text_overlay.py
Comment (sofroniewn) [inline napari/components/text_overlay.py]: See comment above, I don't think we need the return `None` here. I know that was in the code I copied over from the ColorManager, but I think that was to deal with `current_color` when nothing was selected. Here there will always be text that always needs a color. If I've gotten this wrong maybe @kevinyamauchi can weigh in

File: napari/components/text_overlay.py
Comment (lukasz-migas) [reply napari/components/text_overlay.py]: Nope, that makes sense - it's been corrected now.

File: napari/components/text_overlay.py
Comment (kevinyamauchi) [reply napari/components/text_overlay.py]: Sorry for the confusion, @lukasz-migas ! @sofroniewn was right - the `None` that was returned from those color validators was to deal with the case when a color isn't provided.

Comment (lukasz-migas) [reply napari/components/text_overlay.py]: @kevinyamauchi Thanks for clarification, I've removed the if/else from the validator since we always want color.

File: napari/_vispy/vispy_text_visual.py
Comment (kevinyamauchi) [inline napari/_vispy/vispy_text_visual.py]: Should this be 
```suggestion
            transform = [canvas_size[0] - x_offset, y_offset, 0, 0]
```

**Review by kevinyamauchi — COMMENTED** [review]
I think this is pretty much good to go. See my question below. Also, I believe the failed tests are unrelated (see #2618  - I'm submitting a patch to fix this now)

**Review by jni — APPROVED** [review]
Brilliant PR! I'm comfortable with this being in the main space rather than experimental — I think even as we add more capabilities they can be added to a `text_overlay` attribute in a backwards-compatible way, e.g. maybe we support a list of them or just one, etc.

Thanks @lukasz-migas!!! Lots of people are going to get immediate use out of this!

**Comment by kevinyamauchi** [comment]
I just took this for a spin and it's super cool! Nice work, @lukasz-migas !


https://user-images.githubusercontent.com/1120672/116209925-ab62ef80-a742-11eb-81eb-e69bd64aba2d.mov


@jni, if you have time, can you please take a quick look at #2619 (one liner) - I think it fixes an issue unrelated to this PR that are causing the  tests/flake8 to fail.

File: napari/_vispy/vispy_text_visual.py
Comment (sofroniewn) [reply napari/_vispy/vispy_text_visual.py]: Or maybe `x_offset // 2`?

Comment (sofroniewn) [reply napari/_vispy/vispy_text_visual.py]: Looks like `x_offset` actually I think

**Comment by sofroniewn** [comment]
Ok, this looks done now, I've just merged in the master branch to see if all tests will pass - if so will merge into master now. Thanks again!!

**Pull request merged**
